include/miral/miral/idle_listener.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIRAL_IDLE_LISTENER_H
+#define MIRAL_IDLE_LISTENER_H
+
+#include <functional>
+#include <memory>
+
+namespace mir { class Server; }
+
+namespace miral
+{
+
+struct IdleListenerParams
+{
+    using Callback = std::function<void()>;
+    Callback on_dim;
+    Callback on_off;
+    Callback on_wake;
+};
```

Comment:
Given the added complexity of checking these members are set, maybe it would be better to have an interface:
```suggestion
class IdleObserver
{
public:
    IdleObserver() = default;
    virtual ~IdleObserver() = default;
    virtual void on_dim();
    virtual void on_off();
    virtual void on_wake();

private:
    IdleObserver(IdleObserver const&) = delete;
    IdleObserver& operator=(IdleObserver const&) = delete;
};
```
Or at least provide default initialisation (making the contract "don't reset this"):
```suggestion
struct IdleListenerParams
{
    using Callback = std::function<void()>;
    Callback on_dim = []{};
    Callback on_off = []{};
    Callback on_wake = []{};
};
```